### PR TITLE
feat: decouple @bazel/worker from rules_typescript

### PR DESCRIPTION
**Pull request by thesayyn**
## PR Checklist

Please check if your PR fulfills the following requirements:

- [x] Tests for the changes have been added (for bug fixes / features)
- [ ] Docs have been added / updated (for bug fixes / features)

## PR Type

What kind of change does this PR introduce?

- [ ] Bugfix
- [x] Feature (please, look at the "Scope of the project" section in the README.md file)
- [ ] Code style update (formatting, local variables)
- [x] Refactoring (no functional changes, no api changes)
- [ ] Build related changes
- [ ] CI related changes
- [ ] Documentation content changes
- [ ] Other... Please describe:

## What is the current behavior?

`@bazel/worker` uses an old version of bazel worker protocol and coupled with `rules_typescript` which is not maintained anymore.
Since the used protocol is old, it lacks some of the new features such as `cancellation` and `multiplex-workers`.

Issue Number: https://github.com/bazelbuild/rules_nodejs/issues/1306

## What is the new behavior?

- Now the worker_protocol.proto is checked into the third-party directory and the corresponding code is generated by `protoc-gen-ts`. 
- New version of the package does not rely on just-in-time loading of the proto file and gets complete type safety.
- Package is now fully rewritten in typescript and there is no other third-party dependency other than `google-protobuf`.
- It is ready to handle `cancellation` but we have to introduce some breaking changes such as `runOneBuild` returning a cancel callback to handle the cancellation and taking a callback argument to call when the task is complete. similar to what we see on bundlers like webpack and rollup.
- A new argument (third) `console` is passed to `runOneBuild` callback to let the consumer have an API to write debug messages and feel right at home.

## Does this PR introduce a breaking change?

- [ ] Yes
- [x] No

## Other information
the new protocol file is generated by using [protoc-gen-ts](https://github.com/thesayyn/protoc-gen-ts) which is mostly maintained by me and a few open source contributors hence I am not sure if you would like to rely on a third-party project. might be a deal-breaker for you.

additionally, in this PR I have only changed things to make room for the incoming support for multiplex workers.

### CODE REVIEW

**Comment by thesayyn** [comment]
I would like to hear your thoughts and the points that I should improve

**Comment by alexeagle** [comment]
The variations of TS generation for .proto files is pretty confusing for sure. I like the comparison table on the bottom of your readme. https://github.com/thesayyn/protoc-gen-ts#alternatives

However you only use it as a build-time dependency within our repo to generate those files, not a runtime dependency for our users, so the bar is a lot lower on stability. We'll very rarely change the worker_protocol.proto so we'll rarely need to run the tool. If your repo disappears we could find another generator, or worst case we could edit the files by hand if a field or two gets added to the worker_protocol.proto in the future.

The bigger concern to figure out first is if this is a breaking change, then we can't land it yet. I'd rather not start a 5.x branch yet since I'll have to cherry-pick and manage two branches for 6 months until the next major - usually we only start that process a couple months before the release.

So I guess we'd like to explore how to make this non-breaking. Either by restoring back-compat or distributing it alongside the existing implementation.

**Comment by thesayyn** [comment]
Hmm, then I would vote that we make this non-breaking change and make this reflect the current API.  Once we land this, I could send another PR which we could get landed in the next major release.  

I will send another commit to restore backward compatibility.

**Comment by thesayyn** [comment]
Some of the tests are failing. will pick them tomorrow

**Comment by thesayyn** [comment]
I just realized that I have to update a lot of packages to clean up the old way of using the worker package. Which is going to be fun though like meditation.

**Comment by thesayyn** [comment]
Now some packages (those who has worker support) has a direct dependency on `@bazel/worker`.

**Comment by thesayyn** [comment]
@alexeagle I would very much appreciate  get your thoughts on how I could make these packages depend on the worker package and make the tests pass

**Comment by alexeagle** [comment]
I'm on mobile, but check the examples/BUILD file?

**Comment by thesayyn** [comment]
Unfortunately `npm_packages` only replace occurrences in the root package and not dependants of the root package. So when `@bazel/typescript` it will be replaced by `npm_packages` but the dependencies of `@bazel/typescript` which is the dependency of `@bazel/typescript`

**Comment by thesayyn** [comment]
@alexeagle ready for a review. and got all the tests passing

File: examples/BUILD.bazel
Comment (alexeagle) [inline examples/BUILD.bazel:332]: maybe we should set this as the default value for the `resolutions` attribute so we don't have to remember where it has to be set? wouldn't do any harm to include it even for tests that don't require it, right?

File: examples/esbuild/WORKSPACE
Comment (alexeagle) [inline examples/esbuild/WORKSPACE:30]: why switch to yarn? PRs should be minimal to make them easy to review. If you're making a change that looks orthogonal, either put in a separate PR, or anticipate my question and add a comment about why this change is required for your change

File: examples/from_source/package.json
Comment (alexeagle) [inline examples/from_source/package.json:6]: if the example has to change, that means it's a breaking change...

File: packages/typescript/internal/ts_project.bzl
Comment (alexeagle) [inline packages/typescript/internal/ts_project.bzl:692]: nice

File: packages/worker/src/worker_protocol.ts
Comment (alexeagle) [inline packages/worker/src/worker_protocol.ts:2]: if you're going to check in a generated file, we should at least have a test that asserts it's up-to-date

how hard is it to leave the generated file in bazel-out and not check in as a source file?

Comment (thesayyn) [reply packages/worker/src/worker_protocol.ts:2]: there are some reasons behind this

- have IDE recognize the file 
- somehow rules_proto does not work on windows. it throws an error with a message "bazel-out is not an external or internal command" which probably has something to do with how rules_proto tries to invoke protoc. so the windows tests are going to fail. I have seen it before.
- generating this file every time is not an option since ts_project won’t pick it up. it might create cycling dependency between worker and typescript package since they are relying on each other to build (not sure though)

I would prefer to add a generated_file_test and make an assertion.  this test would be disabled on windows though because of the problem I have described above.

File: examples/from_source/package.json
Comment (thesayyn) [reply examples/from_source/package.json:6]: I have no clue how I can make this non-breaking change... may be you have something on your mind.

File: examples/esbuild/WORKSPACE
Comment (thesayyn) [reply examples/esbuild/WORKSPACE:30]: oh, thats because npm ignores the `resolutions` object in the package.json file. so we needed to install the dependencies with yarn_install just like other examples. but if this is a deal-breaker, I will try to find a way to make it work for both yarn_install and npm_install

File: examples/BUILD.bazel
Comment (thesayyn) [reply examples/BUILD.bazel:332]: the only problem I see is having to build this package but bazel is smart enough to use from previous builds so that makes sense to put this as a default. will send a pr for it.

File: examples/from_source/package.json
Comment (thesayyn) [reply examples/from_source/package.json:6]: also, I have no idea why this was needed in the first place. Shouldn’t rules_nodejs list this as a dependency even if the repository is consumed from the source? 

I only added this because the tests were yelling at me

I believe that the problem lies somewhere else.

Comment (thesayyn) [reply examples/from_source/package.json:6]: Shouldn’t this be handled by rules_nodejs_dev_dependencies? perhaps something similar? as I don’t rely on the consumer's npm repository, I don’t see why this should be listed as a dependency by the consumer.

Comment (alexeagle) [reply examples/from_source/package.json:6]: rules_nodejs_dev_dependencies is only for when you want to consume our rules from source. normal users shouldn't need it. but since this is the examples/from_source it actually does make sense here

Comment (thesayyn) [reply examples/from_source/package.json:6]: should I add it to this example? or shall we make separate PR for this?

Comment (alexeagle) [reply examples/from_source/package.json:6]: sorry it was my mistake, as it's building rules_nodejs from source it's expected that our dev dependencies leak through. So far we have the rules_nodejs_dev_dependencies for *starlark* dependencies, but we don't have anything for *npm* dependencies (in theory it should be the root package json but we don't actually publish `@bazel/javascript` we treat it as private)

File: packages/worker/src/metadata.ts
Comment (alexeagle) [inline packages/worker/src/metadata.ts:1]: add a file-level comment explaining what this metadata is

**Review by alexeagle — COMMENTED** [review]
looking really close :)

**Comment by thesayyn** [comment]
- i have the assertion test in place: `//packages/worker:copy_worker_protocol`
- I couldn’t put the update command comment into the generated file. the assertion fails. so instead I put it into the build file. I hope that's fine

File: examples/from_source/package.json
Comment (thesayyn) [reply examples/from_source/package.json:6]: Hmm. That’s what I was expecting. so is this okay to have this change?

File: packages/worker/src/worker_protocol.ts
Comment (thesayyn) [reply packages/worker/src/worker_protocol.ts:2]: have the assertion test in place: //packages/worker:copy_worker_protocol

Comment (thesayyn) [reply packages/worker/src/worker_protocol.ts:2]: I couldn’t put the update command comment into the generated file. the assertion fails. so instead I put it into the build file. I hope that's fine

File: packages/worker/src/metadata.ts
Comment (thesayyn) [reply packages/worker/src/metadata.ts]: I have already changed the file name to `size` and added the documentation to clear up confusion and not invent a slang word in the bazel world.

**Comment by thesayyn** [comment]
I believe that this is ready for another review.

**Review by alexeagle — APPROVED** [review]
super nice 🌮

**Pull request merged**
